Frameworks/CoreGraphics/CGContext.mm

```diff
@@ -1913,71 +1943,155 @@ void CGContextSetPatternPhase(CGContextRef context, CGSize phase) {
+                               return left.run->glyphCount < right.run->glyphCount;
+                           })->run->glyphCount;
+
+    // Shared between transformed glyph runs
+    std::unique_ptr<CGFloat[]> zeroAdvances(new CGFloat[maxGlyphCount]());
+
+    for (size_t i = 0; i < runsCount; ++i) {
+        auto& run = glyphRuns[i].run;
+        if ((fabs(combinedTransform.a * run->fontEmSize) <= c_glyphThreshold &&
+             fabs(combinedTransform.b * run->fontEmSize) <= c_glyphThreshold) ||
+            (fabs(combinedTransform.d * run->fontEmSize) <= c_glyphThreshold &&
+             fabs(combinedTransform.c * run->fontEmSize) <= c_glyphThreshold)) {
+            // Not a failure state! Not drawing the glyphs is *okay*.
+            TraceWarning(TAG, L"Glyphs too small to be rendered");
+        } else {
+            if (fastPathTextDrawing) {
+                runs.emplace_back(glyphRuns[i]);
+            } else {
+                // Get the linear transformation that is the inverse of the text transform
+                // We transform the positions of each glyph by the inverse so they will be positioned correctly
+                // While still being transformed per glyph by the text transformation
+                CGAffineTransform invertedTextTransformation = CGAffineTransformInvert(
+                    CGAffineTransformScale(CGAffineTransformMake(textTransform.a, textTransform.b, textTransform.c, textTransform.d, 0, 0),
+                                           1,
+                                           -1));
+
+                DWRITE_GLYPH_OFFSET* positions = new DWRITE_GLYPH_OFFSET[run->glyphCount];
+                // First glyph's origin is at the given relative position for the glyph run
+                CGPoint runningPosition{ glyphRuns[i].relativePosition.x, std::round(glyphRuns[i].relativePosition.y) };
+                for (size_t j = 0; j < run->glyphCount; ++j) {
+                    // Invert position by text transformation
+                    CGPoint transformedPosition = CGPointApplyAffineTransform(runningPosition, invertedTextTransformation);
+
+                    // Set current glyph's position to be the actual position transformed by the inverted text position
+                    // So when the space is transformed by the text position it will be drawn in the correct real position
+                    positions[j] = DWRITE_GLYPH_OFFSET{ transformedPosition.x + run->glyphOffsets[j].advanceOffset,
+                                                        std::round(transformedPosition.y + run->glyphOffsets[j].ascenderOffset) };
+
+                    // Translate position of next glyph by current glyph's advance
+                    runningPosition.x += run->glyphAdvances[j];
+                }
+
+                DWRITE_GLYPH_RUN* transformedGlyphRun =
+                    new DWRITE_GLYPH_RUN{ run->fontFace,      run->fontEmSize, run->glyphCount, run->glyphIndices,
```

Comment:
std::make_shared this instead of new-ing it. if you don't want to change the contract of GlyphRunData to take a shared_ptr instead of a *, then you can just stuff this in some unrelated vector so that it can get properly scope-released